dvc/checkout.py

```diff
@@ -61,6 +71,80 @@ def _remove(path_info, tree, cache, force=False):
     tree.remove(path_info)
 
 
+def _verify_link(cache, path_info, link_type):
+    if link_type == "hardlink" and cache.tree.getsize(path_info) == 0:
+        return
+
+    if cache.cache_type_confirmed:
+        return
+
+    is_link = getattr(cache.tree, f"is_{link_type}", None)
+    if is_link and not is_link(path_info):
+        cache.tree.remove(path_info)
+        raise DvcException(f"failed to verify {link_type}")
+
+    cache.cache_type_confirmed = True
+
+
+def _do_link(cache, from_info, to_info, link_method):
+    if cache.tree.exists(to_info):
+        raise DvcException(f"Link '{to_info}' already exists!")
+
+    link_method(from_info, to_info)
+
+    logger.debug(
+        "Created '%s': %s -> %s", cache.cache_types[0], from_info, to_info,
+    )
+
+
+@slow_link_guard
+def _try_links(cache, from_info, to_info, link_types):
+    while link_types:
+        link_method = getattr(cache.tree, link_types[0])
+        try:
+            _do_link(cache, from_info, to_info, link_method)
+            _verify_link(cache, to_info, link_types[0])
+            return
+
+        except DvcException as exc:
+            logger.debug(
+                "Cache type '%s' is not supported: %s", link_types[0], exc
+            )
+            del link_types[0]
+
+    raise CacheLinkError([to_info])
+
+
+def _link(cache, from_info, to_info):
+    assert cache.tree.isfile(from_info)
```

Comment:
Wasting a call :( As noted in the PR desc, link() could use some refactoring. Keeping as it was before for now, but will get back to this shortly during optimization.